### PR TITLE
[ENHANCEMENT] Strict Validation of BatchRequest Attributes

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* Strict validation of `BatchRequest` attributes, which users must specify;
* Dedicated `BatchRequestBase` object for use as part of internal `DataSource` / `DataConnector` protocol.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [ENHANCEMENT], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
-
-
-

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

Thank you for submitting!

### CODE REVIEW

File: great_expectations/core/batch.py
Comment (Shinnnyshinshin) [inline great_expectations/core/batch.py:317]: a subtle difference, but an important one :) thanks for checking this

File: great_expectations/core/batch.py
Comment (Shinnnyshinshin) [inline great_expectations/core/batch.py:164]: Thanks for catching this 👍🏼

File: great_expectations/core/batch.py
Comment (Shinnnyshinshin) [inline great_expectations/core/batch.py:172]: @alexsherstinsky this helped with understanding how these changes fit into the larger "picture", but could you tell me a little bit more about the motivation behind this change? 

- what does `BatchRequestBase` enable the code to do which wasn't possible before? 
- what does it simplify? add functionality to?

Comment (alexsherstinsky) [reply great_expectations/core/batch.py:172]: @Shinnnyshinshin : Thanks for this good question -- as `BatchRequestBase` is the crux of this PR.

`BatchRequestBase` is now used for the internal protocol, which @abegong created during the BatchPod.  During that time, the very `BatchRequest` was used for this purpose.  However, while very convenient for internal data interchange, the `BatchRequest` was the same object as THE `BatchRequest` that is part of the user API for getting batches and/or validators.  This was the point of the difficulty.  In the user-accessible-API, `BatchRequest`, MUST insist on the triple, consisting of `data_source_name`, `data_connector_name`, and `data_asset_name`, to be non-NULL -- none of the three can be NULL -- this is the "required non-empty triumvirate".  Whereas for the internal protocol, `BatchRequest` was just used as a flexible bag of attributes, which any fields allowed to be NULL.  Hence, now, `BatchRequestBase` will be the attribute bag for the internal protocol use, whereby NULL values will be allowed, as per the internal needs.  The `BatchRequest` extends `BatchRequestBase` and adds strong validation (as described and more) -- and will now properly validate user input.

Comment (Shinnnyshinshin) [reply great_expectations/core/batch.py:172]: Thank you for the clear explanation :) It makes a lot more sense. I'd like to have this explanation included somewhere.. do you think a version of the above explanation could be added to the docstrings for `BatchRequest` and `BatchRequestBase`? Otherwise it looks good from my end. 

As a related point, I was wondering what you had in mind for the review process. I'd be happy to hit approve on my end, but can hold off if you'd like to have @jcampbell do the final review and approval.

Comment (alexsherstinsky) [reply great_expectations/core/batch.py:172]: @Shinnnyshinshin -- good idea: I added the docstring with the above explanation (with a better phrasing).

I think that approving it now is safe.  There are no changes to the external API -- in fact, it operates more properly.  Internally, the protocol is also cleaner now, whereby it is easier to understand how the attributes are being passed.

So I think this is safe to approve/merge.  Still, just in case, mentioning @jcampbell here (also /cc @abegong).

Thank you!

**Review by Shinnnyshinshin — APPROVED** [review]
Thank you. LGTM :)

**Pull request merged**
